gapir/cc/cached_resource_loader.h

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (C) 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef GAPIR_CACHED_RESOURCE_LOADER_H
+#define GAPIR_CACHED_RESOURCE_LOADER_H
+
+#include "resource_cache.h"
+#include "resource_loader.h"
+
+#include <memory>
+
+namespace gapir {
+
+// CachedResourceLoader implements the ResourceLoader interface. It drives a
+// ResourceCache and requires a fallback loader for fetching uncached resources.
+class CachedResourceLoader : public ResourceLoader {
+ public:
+  static std::unique_ptr<CachedResourceLoader> create(
+      ResourceCache* cache, std::unique_ptr<ResourceLoader> fallbackLoader) {
+    return std::unique_ptr<CachedResourceLoader>(
+        new CachedResourceLoader(cache, std::move(fallbackLoader)));
+  }
+  virtual ~CachedResourceLoader() {}
+
+  // Loads first tries to find the resources in cache and loads the data from
+  // cached if a resource is found. If a resource is not found, use the fallback
+  // loader to fetch the resource data, then load the data to target and put the
+  // data to cache. If the net size of all the resources exceeds size, then
+  // false is returned.
+  virtual bool load(const Resource* resources, size_t count, void* target,
```

Comment:
// Load first tries to find the resources in cache and loads from there
  // if they are found. If a resource is not found it uses the fallback
  // loader to fetch the resource data, loads the data to target and puts the
  // data to cache. If the net size of all the resources exceeds targetSize, then
  // false is returned.